amazon/ion/ioncmodule.c

```diff
@@ -384,48 +412,37 @@ static iERR ionc_write_sequence(hWRITER writer, PyObject* sequence, PyObject* tu
  */
 static iERR ionc_write_struct(hWRITER writer, PyObject* map, PyObject* tuple_as_sexp) {
     iENTER;
-    PyObject * list = PyMapping_Items(map);
-    PyObject * seq = PySequence_Fast(list, "expected a sequence within the map.");
-    PyObject * key = NULL, *val = NULL, *child_obj = NULL;
-    Py_ssize_t len = PySequence_Size(seq);
-    Py_ssize_t i;
-
-    for (i = 0; i < len; i++) {
-        child_obj = PySequence_Fast_GET_ITEM(seq, i);
-        key = PyTuple_GetItem(child_obj, 0);
-        val = PyTuple_GetItem(child_obj, 1);
-        Py_INCREF(child_obj);
-        Py_INCREF(key);
-        Py_INCREF(val);
-
-        if (PyUnicode_Check(key)) {
-            ION_STRING field_name;
-            ion_string_from_py(key, &field_name);
-            IONCHECK(ion_writer_write_field_name(writer, &field_name));
+    PyObject *store = NULL, *key = NULL, *val_list = NULL, *val = NULL;
+    Py_ssize_t pos = 0, i, list_len;
+    if (PyDict_Check(map)) {
```

Comment:
Please ensure we have test coverage for `IonPyDict`, `IonPyStdDict` and std `dict`s.

Reply:
Users should be able to write a standard python `dict` object as an Ion struct, and likewise an `IonPyStdDict` (which they _should_ be able to given that it inherits from `dict`). I do not know if there are existing tests or not.

Reply:
Thanks for the clarification. We do have test for [IonPyDict](https://github.com/amazon-ion/ion-python/blob/d425587d94f8e9fa22368ed1b82a1b65fcb81820/tests/test_simple_types.py#L59) and [dict](https://github.com/amazon-ion/ion-python/blob/d425587d94f8e9fa22368ed1b82a1b65fcb81820/tests/test_simple_types.py#L115). For `IonPyStdDict`, it should be covered in this [test](https://github.com/amazon-ion/ion-python/blob/d425587d94f8e9fa22368ed1b82a1b65fcb81820/tests/test_simple_types.py#L115) since it's subclass of `dict`.